compiler-api/src/main/java/com/squareup/anvil/compiler/api/GeneratedFile.kt

```diff
@@ -8,7 +8,30 @@ import java.io.File
  * @property file the [File] to generate.
  * @property content the file contents to write to [file].
  */
-public data class GeneratedFile(
-  val file: File,
-  val content: String,
+@Deprecated(
+  message = "Use GeneratedFileWithSources instead.",
+  replaceWith = ReplaceWith(
+    expression = "GeneratedFileWithSources(file, content, setOf(TODO(\"Add some source files using <somePsiElement>.containingFileAsJavaFile()\")))",
+    imports = [
+      "com.squareup.anvil.compiler.api.GeneratedFileWithSources",
+      "com.squareup.anvil.compiler.internal.containingFileAsJavaFile",
+    ],
+  ),
+  level = DeprecationLevel.WARNING,
 )
+public data class GeneratedFile(
```

Comment:
What do you think about adding a separate constructor here and deprecating the current no-source-files constructor instead of introducing the new type? That should result in fewer changes for folks migrating their own generators now, as well as later when we would presumably rename `GeneratedFileWithSources` back to just `GeneratedFile` when this original one gets fully deleted (reason being that the inclusion of sources feels odd to have encoded in the name once this current form is gone). From a quick glance that seems like it should also remove the need for the new interface. E.g.
```
public class GeneratedFile {

  public val file: File
  public val content: String
  public val sourceFiles: Set<File>

  public constructor(file: File, content: String, sourceFiles: Set<File>) {
    this.file = file
    this.content = content
    // <sourceFiles assertions go here instead of init>
    this.sourceFiles = sourceFiles
  }

  @Deprecated("stop using this etc")
  public constructor(file: File, content: String) {
    this.file = file
    this.content = content
    this.sourceFiles = emptySet()
  }
}
```

Reply:
I did consider the second constructor, but I wanted a clean break.  That's for two main reasons:

1. `GeneratedFile` is a data class, so binary compatibility is a bit tricky.  It's best to just get away from that.
2. I want to add a warning or exception if the incremental behavior is enabled, but a (custom) generator returns results that don't track anything.  That's technically different than a result that simply doesn't have any source files.